### ISSUE TITLE
thetvdb API v4 to be monetized -- Medusa strategy?

### ISSUE THREAD

**Issue by mmhere** [pymedusa/Medusa#8738]
I guess this is either a feature request or a forward looking support question designed to track thetvdb API changes and impending monetization on their (thetvdb) part.

I just received an email that "v4" API represents their plan to monetize their data.  Projects either have to pay for a project apikey or pass the cost onto end users who will require their own personal apikey (global across projects they use).

I think I'm understanding it properly, but the email is attached below so it can speak for itself.

Deadline:  current APIs go offline in March 2021 according to the email.

---------

Apologies in advance for PDF here;  couldn't figure out how to reasonably preserve formatting and keep it "portable" (so-called).  Didn't want to paste the large block of text here either:

[x.pdf](https://github.com/pymedusa/Medusa/files/5541501/x.pdf)

**Comment by p0psicles** [comment]
Yeah I tried getting in contact, no success yet

**Comment by mmhere** [comment]
> Yeah I tried getting in contact, no success yet

Cool.  Thanks for initiating.

We could have a policy discussion separate from any comms happening with thetvdb.

I'd guess it unlikely that Medusa itself would be able to (or want to) pay for a license for the project.  Who knows, however, whether thetvdb might offer a free apikey at the Medusa project level?  Seems unlikely.

My conclusion, therefore, is that we need to [a] implement against their "v4" API that is supposedly live now, and [b] put in place user-level apikey capture and pass-thru for those that choose to pay individually.

I'm sure this will all go just as smoothly as thetvdb forced-turmoil that began October last yer.  ;-)

**Comment by p0psicles** [comment]
For me it means a. Make sure the alternatives work properly (tvmaze and today) and then b. See what we can do with tvdb changes

**Comment by DAVIZINH0** [comment]
Its posible a workarround to migrate all the series that actually i have with tvdb to the tvmaze indexer?
thanks

**Comment by OmgImAlexis** [comment]
If need be I can setup a proxy for us to use so they only see my key.

**Comment by p0psicles** [comment]
Tnx that's very generous. But, the issue with that is, that then we're getting into a cat&mouse game. Where it's possible they'll start blocking keys. Or IP addresses. And I'd like medusa to be an as much as possible stable app. 

So we need to work something out that's stable. Or we'll look into alternatives

**Comment by OmgImAlexis** [comment]
I’ve messaged them and sent them my contact info and nothing. They’re expecting $6k+ from another group. I feel like no matter what happens we’re not going to be in their good books. 💁‍♀️

**Comment by p0psicles** [comment]
Then we pull the plug when the time comes

**Comment by desepticon** [comment]
I guess one feature that would be nice to implement in the meantime is to be able to change the data provider. Either through a mass update or on a per-episode basis. I didn't see any option for that in the "edit" tab.

**Comment by p0psicles** [comment]
We just refactored add existing show page. I'll use  that to create an indexer change page. It's currently in develop branch if you like a preview

**Comment by desepticon** [comment]
> We just refactored add existing show page. I'll use that to create an indexer change page. It's currently in develop branch if you like a preview

 Thank!

**Comment by mmhere** [comment]
Do folks think keeping thetvdb as an option (with users putting in their own apikey, whether they get that by paying money for it, or by being gifted an apikey because they add data to thetvdb sufficiently and frequently enough[*]).

If it is not too much additional developement work here to track their "v4" API changes (sure to be ongoing, capricious, and arbitrary [and broken frequently]), it seems like it would be nice to keep access to their data as an option.

Along the lines of which provider to use -- how do people feel about tvmaze?  Is it as fully populated with complete data as is thetvdb?  Is it updated frequently enough?  Is its API access any more stable? Given the many recent downtime problems with thetvdb and API breakages [bad/incomplete data served] over the past year, perhaps tvmaze is a better option overall if it would actually reliably serve [some] data?

I'm willing to switch entirely to tvmaze if it is up to snuff.  This is, however, a major undertaking given that show meta data is used by other platforms (various players) as well as here with Medusa.  I'd like to keep the retrievers and players in synch.  If changed in Medusa the effort to also change in players is painful.

[*]The communique I quoted above in PDF claims that thetvdb will have a policy (not yet fully defined apparently) that allows users who submit data to be given a valid apikey.  It sounds like this is one way to keep access to their data active without having to pay for it using legal tender, but instead pay for with a sort of work trade.

**Comment by medariox** [comment]
We haven't decided yet. As a FOSS project the only thing that would make sense is to drop support completely. Their behavior in this situation has been despicable at best.

We have added support for TVmaze and TMDB since a long time and so far haven't received any complaints about either of them. Why don't you try them out with a few shows?

**Comment by mmhere** [comment]
Yes, I've started using TVmaze so it is good to hear that people haven't had issue with it.

Switching everything over is, as mentioned, a significant effort.

Along those lines it is very cool to see the migrate logic being developed (mentioned above) which will mitigate the pain of changing everything in Medusa.

Migrating in other [unnamed] projects/players poses a different challenge

---

> Their behavior in this situation has been despicable at best.

Agreed.  Keeping it FOSS (meaning use of TVmaze) is certainly preferred.

**Comment by mmhere** [comment]
Do we feel that the despicable behaviour precludes continuing to provide support for thetvdb?

Are there other examples of info providers (be it show meta data or other stuff, I dunno, subtitles) who force access thru non-FOSS mechanisms, which we still support?

**Comment by p0psicles** [comment]
For subtitle providers we do support supplying an api key to use them. But those are part of Medusa core functionality. One thing to consider before dropping tvdb support is that anime users cannot make use of alternatives like tvmaze and tmdb, because thexem.de currently does not use them. Well, there are also some non-anime shows that depend on thexem.de. So that's something we should look out for. 

But I'd prefer if tvmaze or tmdb would add support for alternative season/episode numbering. So that a user can pick it when adding a show. Like select from indexer default, Netflix, scene. But thats more like a personal wish

**Comment by mmhere** [comment]
As far as implementing changes here to track the presumed-to-be moving target that is the "v4" thetvdb API (assuming we don't drop support), I'm willing to QA it.  I'm a SWdev IRL so I think I could be useful.  I plan to get a user apikey whether paid for by work trade or cash.

How likely are we to actually target the "v4" API then?  Anyone have a feel for this yet?  I certainly don't want to create extra dev work for functionality few will use if that is the case.

It does, however, seem quite unlikely that we would implement a project-level apikey for Medusa:

1. Seems highly unlikely they'd simply give this to us.
2. We surely aren't going to pay thousands of duckets for it as a project, right?

**Comment by p0psicles** [comment]
@thezoggy what are your thoughts on this?

**Comment by p0psicles** [comment]
Also started a discussion here. Tvmaze is very fast with feedback. https://www.tvmaze.com/threads/4515/alternative-episode-numbering#62122

@OmgImAlexis also inviting you to the discussion

**Comment by thezoggy** [comment]
TL;DR; It is still a little too early to tell. The new v4 api is not yet feature complete to replace v3 and with tvdb's sorrid history of api migrations I cant see it going smoothly or on time.

Personally, if they would have asked for donations on v1 days back before the v2+ mess I would have been fine to donate even after all the amount of time I've spent contributing to the site. The v2+ stuff turned me off. So for right now I would not recommend anyone giving them money until v4 is proven not to be a shitshow.

kodi/plex have giant user bases, they probably take up the bulk of tvdb api load. users of these type of apps dont really benefit tvdb as they never go to the website (so no chance to track user data / get ad views). but these apps have sponsors and other things that could help work out a deal and keep the user experience the same by not passing it down to the user. kodi for awhile now stopped defaulting to tvdb and moved over to tmdb for tvshows.. so might just see more people jump ship. also, someone would need to spend time on making a v4 compatible scraper for python2 (kodi 18 and older) and one for python3 (kodi 19)..  so that also adds some wrinkle onto who is going to spend time to do it.

sonarr for example has skyhook where all their users actually hit (skyhook gets data from tvdb, caches it and allows sonarr devs to clean up the data and do some logic on it). So really sonarr user base is reduced to minimal impact to tvdb because of this. also with them running the proxy that means they could do a key there but of course the one key for many is a no-no and why negotiations has to happen so it doesnt get abused. if you were to try and replicate this on the down low.. im sure your ip would get banned and so on.

**Comment by thezoggy** [comment]
i had started to add tvmaze at one point (back when tvrage was around) but didnt because their data was so bad (they were pretty new of a site with low user base). i'll have to poke around again.

**Comment by nydave69** [comment]
No love for TheTVDB here.....

Just need an easy way to port over to tvmaze...

**Comment by Rouzax** [comment]
Just want to link my old feature request here 😄 
https://github.com/pymedusa/Medusa/issues/6531
